File: addTwoNumbers.c
struct ListNode {
         int val;
         struct ListNode *next;
};

struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    struct ListNode *sumList = malloc(sizeof(struct ListNode));
    struct ListNode *pt = sumList;
    int sum = 0;
    int ten = 0;
    while (l1 || l2 || ten > 0) {
        int val1 = l1 ? l1->val : 0;
        int val2 = l2 ? l2->val : 0;
        sum = val1 + val2 + ten;
        ten = 0;
        if (sum >= 10) {
            ten = 1;
            sum -= 10;
        }
        struct ListNode *newNode = malloc(sizeof(struct ListNode));
        newNode->val = sum;
        pt->next = newNode;
        if (l1) {
            l1 = l1->next ? l1->next : NULL;
        }
        if (l2) {
            l2 = l2->next ? l2->next : NULL;
        }
        pt = newNode;
    }
    return sumList->next;
}

File: leetcode/leetcode/Solutions/LinkedList/AddTwoNumbers.swift
//
//  AddTwoNumbers.swift
//  leetcode
//
//  Created by Wang Kai on 2018/3/22.
//  Copyright © 2018年 Pirate. All rights reserved.
//

import Foundation

class AddTwoNumbers {
    func addTwoNumbers(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
        guard l1 != nil else {
            return l2
        }
        guard l2 != nil else {
            return l1
        }
        var l1 = l1
        var l2 = l2
        let header = ListNode(0)
        var node = header
        var carry = 0
        var sum = 0
        while l1 != nil || l2 != nil || carry > 0 {
            let val1 = l1?.val ?? 0
            let val2 = l2?.val ?? 0
            sum = val1 + val2 + carry
            carry = 0
            if sum >= 10 {
                carry = 1
                sum -= 10
            }
            let newNode = ListNode.init(sum)
            node.next = newNode
            l1 = l1?.next
            l2 = l2?.next
            node = newNode
        }

        return header.next
    }
}
